RFC-0004-expression-eval.md

```diff
@@ -0,0 +1,80 @@
+
+1) It requires a forked set of functions to be written in Java, which is a maintenance burden.
+2) An unfenced function written in C++ must have a corresponding Java implementation in order to work properly.
+
+RFC-0003 proposed a mechanism to both add a function registry to allow C++ implemented functions to be exposed to the coordinator without a corresponding Java implementation, and to add constant folding support via remote function evaluation.  This RFC has a similar goal, but represents an improvement over that approach.
+
+In this RFC, we propose to optimize row expressions as a whole, not just function calls.  This has a few advantages:
+
+1) It allows for constant folding to be applied to expressions which are not function calls, such as special form expressions.
+2) It can reduce the overhead of constant folding by allowing the evaluation engine to evaluate the expression as a whole, rather than evaluating each function call individually.
+
+### [Optional] Goals
+
+* Allow for a pluggable implementation of expression evaluation and use this in the Presto optimizer for constant folding
+* Allow connector authors to use this plugin to optimize their own expressions (for example, for filter pushdown)
+
+### [Optional] Non-goals
+
+## Proposed Implementation
+
+A new SPI will be added to Presto to allow for a pluggable implementation of expression evaluation.  A default implementation of this will be added which delegates to the existing expresion evaluation code, but this may be overridden by a custom plugin.
+
+A new plugin will be implemented which will optimize row expressions by serializing them and sending them to the Presto sidecar for evaluation.  The Presto sidecar will have a new endpoint which can take in a serialized row expression and return the result of evaluating it.
+
+> Endpoint: /v1/evaluate
+> 
+> HTTP verb: POST
+> 
+>Body: JSON list of serialized row expressions
+>
+> Response: JSON list of optimized serialized row expressions
+ 
+The new plugin will be integrated into the Presto optimizer through a new optimizer rule similar to `RowExpressionOptimizer`, except which uses the new SPI to evaluate the expressions.  There will be a feature toggle which can enable or disable the use of this new SPI, however eventually it will become mandatory to use it.
+
+## [Optional] Metrics
+
+This is a 0 to 1 feature, and its success will be largely dictated by two metrics:
+
+1) The performance of evaluating expressions, particularly large, complicated expressions.  This can be measured by comparing the time it takes to evaluate a query with and without the new SPI.
```

Comment:
Actually, it might be more important to measure overhead of constant-folding simple expressions like 1 + 2.

Reply:
Agreed, I will add a metric that also mentions the overhead for simple expressions, as we'd want to prove it's negligible.